boot/model_test.go

```diff
@@ -0,0 +1,1069 @@
+			})
+		default:
+			c.Errorf("unexpected additional call to secboot.ResealKeys (call # %d)", resealKeysCalls)
+		}
+		// what's in params?
+		switch resealKeysCalls {
+		case 1:
+			c.Assert(params.ModelParams, HasLen, 2)
+			c.Assert(params.ModelParams[0].Model.Model(), Equals, "my-model-uc20")
+			c.Assert(params.ModelParams[1].Model.Model(), Equals, "my-new-model-uc20")
+		case 2:
+			// recovery key resealed for current model only
+			c.Assert(params.ModelParams, HasLen, 1)
+			c.Assert(params.ModelParams[0].Model.Model(), Equals, "my-model-uc20")
+		case 3, 4:
+			// try model has become current
+			c.Assert(params.ModelParams, HasLen, 1)
+			c.Assert(params.ModelParams[0].Model.Model(), Equals, "my-new-model-uc20")
+		}
+		// what's in modeenv?
+		switch resealKeysCalls {
+		case 1, 2:
+			// keys are first resealed for both models
+			c.Assert(currForSealing, Equals, "canonical/my-model-uc20,dangerous,"+s.keyID)
+			c.Assert(tryForSealing, Equals, "canonical/my-new-model-uc20,secured,"+s.keyID)
+			// boot/device/model is still the old file
+			c.Assert(filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"), testutil.FileContains,
+				"model: my-model-uc20\n")
+			c.Assert(filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"), testutil.FileContains,
+				"grade: dangerous\n")
+		case 3:
+			// and finally just for the new model
+			c.Assert(currForSealing, Equals, "canonical/my-new-model-uc20,secured,"+s.keyID)
+			c.Assert(tryForSealing, Equals, "/,,")
+			// boot/device/model is the new model by this time
+			c.Assert(filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"), testutil.FileContains,
+				"model: my-new-model-uc20\n")
+			c.Assert(filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"), testutil.FileContains,
+				"grade: secured\n")
+		}
+
+		if resealKeysCalls == 3 {
+			return fmt.Errorf("fail on second try")
+		}
+
+		return nil
+	})
+	defer restore()
+
+	err = boot.DeviceChange(s.oldUc20dev, s.newUc20dev)
+	c.Assert(err, ErrorMatches, `cannot reseal the encryption key: fail on second try`)
+	c.Assert(resealKeysCalls, Equals, 3)
+	// old model file was restored
+	c.Check(filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"), testutil.FileContains,
+		"model: my-model-uc20\n")
+
+	m, err := boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	currForSealing := boot.ModelUniqueID(m.ModelForSealing())
+	tryForSealing := boot.ModelUniqueID(m.TryModelForSealing())
+	c.Assert(currForSealing, Equals, "canonical/my-model-uc20,dangerous,"+s.keyID)
+	// try model has been cleared
+	c.Assert(tryForSealing, Equals, "/,,")
+}
+
+func (s *modelSuite) TestDeviceChangeRebootBeforeNewModel(c *C) {
+	// system is encrypted
+	s.stampSealedKeys(c, s.rootdir)
+
+	// mangle the modeenv to observe the right model content restored for
+	// resealing
+	m, err := boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	m.Model = "foo"
+	m.TryModel = "bar"
+	m.TryBrandID = "baz"
```

Comment:
this kind of test tweak is a bit strange to me, the tests are mostly trying to be realistic but this bit is not oviously so.

My original thinking was to have a test where only the ModelSignKeyID changes and we get back to this device change after rebooting after the modeenv Write at line 99...